cdap-app-fabric/src/test/java/co/cask/cdap/scheduler/CoreSchedulerServiceTest.java

```diff
@@ -320,6 +339,63 @@ public boolean apply(Job job) {
     testScheduleUpdate("delete");
   }
 
+  @Test
+  @Category(XSlowTests.class)
+  public void testProgramEvents() throws Exception {
+    // Deploy the app
+    deploy(AppWithMultipleSchedules.class);
+    CConfiguration cConf = CConfiguration.create();
+
+    TopicId programEventTopic = NamespaceId.SYSTEM.topic(cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC));
+    ProgramStateWriter programStateWriter = new MessagingProgramStateWriter(cConf, messagingService);
+
+    long lastProcessed = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+
+    // These notifications should not trigger the program
+    ProgramRunId anotherWorkflowRun = ANOTHER_WORKFLOW.run(RunIds.generate());
+    programStateWriter.start(anotherWorkflowRun, new SimpleProgramOptions(anotherWorkflowRun.getParent()), null);
+    programStateWriter.running(anotherWorkflowRun, null);
+    programStateWriter.error(anotherWorkflowRun, null);
+    waitUntilProcessed(programEventTopic, lastProcessed);
+    Assert.assertEquals(0, getRuns(TRIGGERED_WORKFLOW));
+
+    lastProcessed = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+    ProgramRunId someWorkflowRun = SOME_WORKFLOW.run(RunIds.generate());
+    programStateWriter.start(someWorkflowRun, new SimpleProgramOptions(someWorkflowRun.getParent()), null);
+    programStateWriter.running(someWorkflowRun, null);
+    programStateWriter.killed(someWorkflowRun);
+    waitUntilProcessed(programEventTopic, lastProcessed);
+    Assert.assertEquals(0, getRuns(TRIGGERED_WORKFLOW));
+
+    // Enable the schedule
+    scheduler.enableSchedule(APP_MULT_ID.schedule(AppWithMultipleSchedules.WORKFLOW_COMPLETED_SCHEDULE));
+
+    String dummyUserKey = "dummy.user.argument.key";
+    String dummyUserValue = "dummy.user.argument.value";
+    // Start a program that writes to the workflow token with some user arguments
+    startProgram(ANOTHER_WORKFLOW, ImmutableMap.of(dummyUserKey, dummyUserValue), 200);
+    waitForCompleteRuns(1, ANOTHER_WORKFLOW);
+
+    // Wait for a notification to come to the scheduler
+    waitUntilProcessed(programEventTopic, lastProcessed);
```

Comment:
this wait is not needed since `assertProgramRuns(TRIGGERED_WORKFLOW, ProgramRunStatus.COMPLETED, 1);` is sufficient. Also we can't determine what's the time to use here since we don't know when completed status of `ANOTHER_WORKFLOW` is published